lib/facter/udev.rb

```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+# Fact: udev
+#
+# Purpose:
+#   Provide access to udev db data.
+#
+# Resolution:
+#   Check if the kernel is linux and that the udevadm command is present.
+#
+# Caveats:
+#   Devices without a /dev/<foo> link are filtered out.
+#
+Facter.add(:udev) do
+  confine kernel: :linux
+  confine(:udevadm) { |x| x.key?('path') }
+
+  def parse_dev(str)
+    d = {}
+
+    str.split(%r{\n}).each do |l|
+      (type, data) = l.split(': ', 2)
+      # See the source for a list of letter codes:
+      # https://github.com/systemd/systemd/blob/a7b2aa658f35f4b9e91915eaa72afa648d0f9119/src/udev/udevadm-info.c#L182
+      case type
+      when 'N'
+        d[:name] = data
+      when 'P'
+        d[:path] = data
+      when 'S'
+        d[:symlink].nil? ? d[:symlink] = [data] : d[:symlink].push(data)
+      when 'E'
+        (key, value) = data.split('=', 2)
+        d[:property].nil? ? d[:property] = { key => value } : d[:property][key] = value
```

Comment:
This may be more accurate:
```suggestion
        d.key?(:property) ? d[:property] = { key => value } : d[:property][key] = value
```

Another alternative is to initialize `d` as:
```ruby
d = {
  symlink: [],
  property: {},
}
```

Then this and the symlink code above is greatly simplified.

Reply:
I agree that `#key?` is more correct. I'm unsure if initializing the hash keys is the right approach since not all devices will have a symlink.

Reply:
You could argue that an empty list also signifies that there are no symlinks.